### PR TITLE
Hotfix for breakdowns in GPMR

### LINKED ISSUE #466: NaN with GPMR on certain systems
[bug gpmr.zip](https://github.com/JuliaSmoothOptimizers/Krylov.jl/files/7677986/bug.gpmr.zip)

Hi, I have an issue with GPMR when solving some systems (I provided one in the above zip file with the code to read it):

```julia
using SparseArrays, LinearAlgebra, Krylov, DelimitedFiles, MatrixMarket
read_path = raw"C:\Users\Geoffroy Leconte\Documents\doctorat\code\bugs\bug gpmr"
A = MatrixMarket.mmread(string(read_path, "/A.mtx"))
E = readdlm(string(read_path, "/E.rhs"), Float64)[:]
δ = readdlm(string(read_path, "/delta.rhs"), Float64)[:][1]
ξ1 = readdlm(string(read_path, "/xi1.rhs"), Float64)[:]
ξ2 = readdlm(string(read_path, "/xi2.rhs"), Float64)[:]

(x, y, stats) = gpmr(A', A, ξ1, ξ2, 
                      C = inv(Diagonal(sqrt.(E))), E = inv(Diagonal(sqrt.(E))), 
                      D = (1.0 / sqrt(δ)) .* I,  F = (1.0 / sqrt(δ)) .* I, λ = -1.0)
```
This leads to `NaN`. I don't know if it is supposed to fail on such a system but if it is, it would be convenient to have something that terminates the algorithm and returns an error.

Since this is a saddle-point system I can solve it with TRIMR or TRICG (and it works):
```julia
(x, y, stats) = tricg(A', ξ1, ξ2, M = inv(Diagonal(E)), N = (1.0 / δ) .* I, flip=true)
```

### PR DESCRIPTION
**Pull request by amontoison**
close #466 

@dpo 
Our orthogonal Hessenberg process continues if `hₖ₊₁.ₖ = 0` or `fₖ₊₁.ₖ = 0`.
We just need to not normalize `vₖ₊₁` or `uₖ₊₁` if we encounter these cases.

In our article, we propose to handle breakdowns theoretically like this
```
"""In case of breakdown, which happens if Auₖ ∈ Span{v₁ , ... , vₖ} or Bvₖ ∈ Span{u₁ , ... , uₖ},
we choose an arbitrary unit vₖ₊₁ ⊥ Span{v₁, ... , vₖ} or uₖ₊₁ ⊥ Span{u₁, ... , uₖ}
and set βₖ₊₁ = 0 or γₖ₊₁ = 0, respectively."""
```
in the proof of the orthogonal Hessenberg process (GPMR paper).
For this reason,  the process should be stopped after `p := min{m,n}` iterations.

But numerically, it also works to set `βₖ₊₁ = vₖ₊₁ = 0` if `Auₖ ∈ Span{v₁ , ... , vₖ}` or `γₖ₊₁ = uₖ₊₁ = 0` if `Bvₖ ∈ Span{u₁ , ... , uₖ}`.
We can continue the process until `max{m,n}`.
At `max{m,n}`, in the worst case, `hₖ₊₁.ₖ = fₖ₊₁.ₖ = 0` and the exact solution is found.
It's the missing piece to prove the convergence of GPMR !

### CODE REVIEW

**Comment by amontoison** [comment]
@geoffroyleconte 
With the linear system of #466 : 
```
GPMR: system of 9806 equations in 9806 variables
    k     ‖rₖ‖   hₖ₊₁.ₖ   fₖ₊₁.ₖ
    0  3.7e+03  ✗ ✗ ✗ ✗  ✗ ✗ ✗ ✗
    1  2.4e+03  1.7e+02  0.0e+00
    2  1.1e+03  0.0e+00  8.5e+01
    3  1.1e+03  1.4e+02  0.0e+00
    4  7.8e+02  0.0e+00  1.2e+02
    5  7.8e+02  1.1e+02  0.0e+00
    6  6.7e+02  0.0e+00  1.3e+02
    7  6.7e+02  9.9e+01  0.0e+00
    8  6.1e+02  0.0e+00  1.2e+02
    9  6.1e+02  1.2e+02  0.0e+00
   10  5.7e+02  0.0e+00  1.2e+02
   11  5.7e+02  1.1e+02  0.0e+00
   12  5.4e+02  0.0e+00  1.3e+02
   13  5.4e+02  1.0e+02  0.0e+00
   14  5.3e+02  0.0e+00  1.3e+02
   15  5.3e+02  1.1e+02  0.0e+00
   16  5.2e+02  0.0e+00  1.2e+02
   17  5.2e+02  1.1e+02  0.0e+00
   18  5.1e+02  0.0e+00  1.2e+02
   19  5.1e+02  1.2e+02  0.0e+00
   20  5.0e+02  0.0e+00  1.1e+02
   21  5.0e+02  1.2e+02  0.0e+00
   22  4.9e+02  0.0e+00  1.2e+02
   23  4.9e+02  1.1e+02  0.0e+00
   24  4.8e+02  0.0e+00  1.2e+02
   25  4.8e+02  1.2e+02  0.0e+00
   26  4.7e+02  0.0e+00  1.2e+02
   27  4.7e+02  1.1e+02  0.0e+00
   28  4.7e+02  0.0e+00  1.3e+02
   29  4.7e+02  1.1e+02  0.0e+00
   30  4.6e+02  0.0e+00  1.2e+02
   31  4.6e+02  1.2e+02  0.0e+00
   32  4.6e+02  0.0e+00  1.1e+02
   33  4.6e+02  1.2e+02  0.0e+00
   34  4.5e+02  0.0e+00  1.1e+02
   35  4.5e+02  1.2e+02  0.0e+00
   36  4.5e+02  0.0e+00  1.1e+02
   37  4.5e+02  1.2e+02  0.0e+00
   38  4.4e+02  0.0e+00  1.1e+02
   39  4.4e+02  1.2e+02  0.0e+00
   40  4.4e+02  0.0e+00  1.1e+02
   41  4.4e+02  1.2e+02  0.0e+00
   42  4.3e+02  0.0e+00  1.1e+02
   43  4.3e+02  1.3e+02  0.0e+00
   44  4.2e+02  0.0e+00  1.1e+02
   45  4.2e+02  1.1e+02  0.0e+00
   46  4.2e+02  0.0e+00  1.2e+02
   47  4.2e+02  1.2e+02  0.0e+00
   48  4.1e+02  0.0e+00  1.2e+02
   49  4.1e+02  1.1e+02  0.0e+00
   50  4.0e+02  0.0e+00  1.2e+02
   51  4.0e+02  1.2e+02  0.0e+00
   52  4.0e+02  0.0e+00  1.1e+02
   53  4.0e+02  1.2e+02  0.0e+00
   54  3.9e+02  0.0e+00  1.2e+02
   55  3.9e+02  1.1e+02  0.0e+00
   56  3.9e+02  0.0e+00  1.2e+02
   57  3.9e+02  1.2e+02  0.0e+00
   58  3.8e+02  0.0e+00  1.1e+02
   59  3.8e+02  1.3e+02  0.0e+00
   60  3.7e+02  0.0e+00  1.1e+02
   61  3.7e+02  1.2e+02  0.0e+00
   62  3.6e+02  0.0e+00  1.1e+02
   63  3.6e+02  1.2e+02  0.0e+00
   64  3.6e+02  0.0e+00  1.2e+02
   65  3.6e+02  1.1e+02  0.0e+00
   66  3.5e+02  0.0e+00  1.1e+02
   67  3.5e+02  1.2e+02  0.0e+00
   68  3.4e+02  0.0e+00  1.1e+02
   69  3.4e+02  1.2e+02  0.0e+00
   70  3.4e+02  0.0e+00  1.1e+02
   71  3.3e+02  1.2e+02  0.0e+00
   72  3.3e+02  0.0e+00  1.1e+02
   73  3.3e+02  1.1e+02  0.0e+00
   74  3.2e+02  0.0e+00  1.3e+02
   75  3.2e+02  1.1e+02  0.0e+00
   76  3.2e+02  0.0e+00  1.3e+02
   77  3.1e+02  1.1e+02  0.0e+00
   78  3.1e+02  0.0e+00  1.2e+02
   79  3.1e+02  1.2e+02  0.0e+00
   80  3.1e+02  0.0e+00  1.1e+02
   81  3.1e+02  1.2e+02  0.0e+00
   82  3.0e+02  0.0e+00  1.1e+02
   83  3.0e+02  1.2e+02  0.0e+00
   84  3.0e+02  0.0e+00  1.2e+02
   85  3.0e+02  1.1e+02  0.0e+00
   86  2.9e+02  0.0e+00  1.2e+02
   87  2.9e+02  1.1e+02  0.0e+00
   88  2.9e+02  0.0e+00  1.2e+02
   89  2.9e+02  1.2e+02  0.0e+00
   90  2.8e+02  0.0e+00  1.2e+02
   91  2.8e+02  1.1e+02  0.0e+00
   92  2.8e+02  0.0e+00  1.2e+02
   93  2.8e+02  1.1e+02  0.0e+00
   94  2.8e+02  0.0e+00  1.3e+02
   95  2.8e+02  1.1e+02  0.0e+00
   96  2.8e+02  0.0e+00  1.3e+02
   97  2.7e+02  1.0e+02  0.0e+00
   98  2.7e+02  0.0e+00  1.3e+02
   99  2.7e+02  1.1e+02  0.0e+00
  100  2.7e+02  0.0e+00  1.3e+02
  101  2.7e+02  1.1e+02  0.0e+00
  102  2.7e+02  0.0e+00  1.2e+02
  103  2.7e+02  1.1e+02  0.0e+00
  104  2.6e+02  0.0e+00  1.2e+02
  105  2.6e+02  1.1e+02  0.0e+00
  106  2.6e+02  0.0e+00  1.2e+02
  107  2.6e+02  1.2e+02  0.0e+00
  108  2.6e+02  0.0e+00  1.1e+02
  109  2.6e+02  1.2e+02  0.0e+00
  110  2.6e+02  0.0e+00  1.1e+02
  111  2.5e+02  1.3e+02  0.0e+00
  112  2.5e+02  0.0e+00  1.0e+02
  113  2.5e+02  1.3e+02  0.0e+00
  114  2.5e+02  0.0e+00  1.1e+02
  115  2.5e+02  1.2e+02  0.0e+00
  116  2.5e+02  0.0e+00  1.1e+02
  117  2.5e+02  1.2e+02  0.0e+00
  118  2.5e+02  0.0e+00  1.1e+02
  119  2.5e+02  1.2e+02  0.0e+00
  120  2.4e+02  0.0e+00  1.2e+02
  121  2.4e+02  1.0e+02  0.0e+00
  122  2.4e+02  0.0e+00  1.3e+02
  123  2.4e+02  1.1e+02  0.0e+00
  124  2.4e+02  0.0e+00  1.3e+02
  125  2.4e+02  1.1e+02  0.0e+00
  126  2.4e+02  0.0e+00  1.2e+02
  127  2.4e+02  1.1e+02  0.0e+00
  128  2.3e+02  0.0e+00  1.3e+02
  129  2.3e+02  9.9e+01  0.0e+00
  130  2.3e+02  0.0e+00  1.3e+02
  131  2.3e+02  1.1e+02  0.0e+00
  132  2.3e+02  0.0e+00  1.2e+02
  133  2.2e+02  1.1e+02  0.0e+00
  134  2.2e+02  0.0e+00  1.2e+02
  135  2.1e+02  1.1e+02  0.0e+00
  136  2.1e+02  0.0e+00  1.3e+02
  137  2.0e+02  1.1e+02  0.0e+00
  138  2.0e+02  0.0e+00  1.2e+02
  139  2.0e+02  1.1e+02  0.0e+00
  140  2.0e+02  0.0e+00  1.2e+02
  141  1.9e+02  1.2e+02  0.0e+00
  142  1.9e+02  0.0e+00  1.1e+02
  143  1.8e+02  1.3e+02  0.0e+00
  144  1.8e+02  0.0e+00  1.1e+02
  145  1.8e+02  1.3e+02  0.0e+00
  146  1.8e+02  0.0e+00  1.0e+02
  147  1.8e+02  1.3e+02  0.0e+00
  148  1.8e+02  0.0e+00  1.1e+02
  149  1.8e+02  1.2e+02  0.0e+00
  150  1.7e+02  0.0e+00  1.2e+02
  151  1.7e+02  1.2e+02  0.0e+00
  152  1.7e+02  0.0e+00  1.2e+02
  153  1.7e+02  1.2e+02  0.0e+00
  154  1.7e+02  0.0e+00  1.1e+02
  155  1.7e+02  1.2e+02  0.0e+00
  156  1.7e+02  0.0e+00  1.1e+02
  157  1.7e+02  1.2e+02  0.0e+00
  158  1.7e+02  0.0e+00  1.1e+02
  159  1.7e+02  1.2e+02  0.0e+00
  160  1.7e+02  0.0e+00  1.1e+02
  161  1.7e+02  1.2e+02  0.0e+00
  162  1.6e+02  0.0e+00  1.1e+02
  163  1.6e+02  1.2e+02  0.0e+00
  164  1.6e+02  0.0e+00  1.1e+02
  165  1.6e+02  1.3e+02  0.0e+00
  166  1.6e+02  0.0e+00  1.1e+02
  167  1.6e+02  1.2e+02  0.0e+00
  168  1.6e+02  0.0e+00  1.2e+02
  169  1.6e+02  1.1e+02  0.0e+00
  170  1.6e+02  0.0e+00  1.2e+02
  171  1.6e+02  1.0e+02  0.0e+00
  172  1.6e+02  0.0e+00  1.3e+02
  173  1.5e+02  1.0e+02  0.0e+00
  174  1.5e+02  0.0e+00  1.3e+02
  175  1.5e+02  9.7e+01  0.0e+00
  176  1.5e+02  0.0e+00  1.3e+02
  177  1.5e+02  1.0e+02  0.0e+00
  178  1.5e+02  0.0e+00  1.3e+02
  179  1.5e+02  1.0e+02  0.0e+00
  180  1.5e+02  0.0e+00  1.2e+02
  181  1.4e+02  1.1e+02  0.0e+00
  182  1.4e+02  0.0e+00  1.2e+02
  183  1.4e+02  1.1e+02  0.0e+00
  184  1.4e+02  0.0e+00  1.2e+02
  185  1.4e+02  1.2e+02  0.0e+00
  186  1.4e+02  0.0e+00  1.2e+02
  187  1.3e+02  1.1e+02  0.0e+00
  188  1.3e+02  0.0e+00  1.2e+02
  189  1.3e+02  1.1e+02  0.0e+00
  190  1.3e+02  0.0e+00  1.2e+02
  191  1.3e+02  1.2e+02  0.0e+00
  192  1.3e+02  0.0e+00  1.1e+02
  193  1.3e+02  1.2e+02  0.0e+00
  194  1.2e+02  0.0e+00  1.1e+02
  195  1.2e+02  1.2e+02  0.0e+00
  196  1.2e+02  0.0e+00  1.1e+02
  197  1.2e+02  1.2e+02  0.0e+00
  198  1.2e+02  0.0e+00  1.1e+02
  199  1.2e+02  1.2e+02  0.0e+00
  200  1.2e+02  0.0e+00  1.2e+02
  201  1.2e+02  1.1e+02  0.0e+00
  202  1.2e+02  0.0e+00  1.2e+02
  203  1.2e+02  1.1e+02  0.0e+00
  204  1.2e+02  0.0e+00  1.2e+02
  205  1.1e+02  1.1e+02  0.0e+00
  206  1.1e+02  0.0e+00  1.2e+02
  207  1.1e+02  1.2e+02  0.0e+00
  208  1.1e+02  0.0e+00  1.2e+02
  209  1.1e+02  1.1e+02  0.0e+00
  210  1.1e+02  0.0e+00  1.2e+02
  211  1.1e+02  1.1e+02  0.0e+00
  212  1.1e+02  0.0e+00  1.2e+02
  213  1.0e+02  1.2e+02  0.0e+00
  214  1.0e+02  0.0e+00  1.1e+02
  215  1.0e+02  1.2e+02  0.0e+00
  216  1.0e+02  0.0e+00  1.1e+02
  217  1.0e+02  1.3e+02  0.0e+00
  218  1.0e+02  0.0e+00  1.1e+02
  219  1.0e+02  1.3e+02  0.0e+00
  220  1.0e+02  0.0e+00  1.0e+02
  221  1.0e+02  1.4e+02  0.0e+00
  222  9.9e+01  0.0e+00  9.9e+01
  223  9.9e+01  1.4e+02  0.0e+00
  224  9.8e+01  0.0e+00  1.0e+02
  225  9.8e+01  1.3e+02  0.0e+00
  226  9.6e+01  0.0e+00  1.0e+02
  227  9.6e+01  1.4e+02  0.0e+00
  228  9.2e+01  0.0e+00  9.5e+01
  229  9.2e+01  1.4e+02  0.0e+00
  230  8.7e+01  0.0e+00  1.0e+02
  231  8.7e+01  1.3e+02  0.0e+00
  232  7.8e+01  0.0e+00  9.7e+01
  233  7.8e+01  1.4e+02  0.0e+00
  234  6.7e+01  0.0e+00  9.8e+01
  235  6.7e+01  1.4e+02  0.0e+00
  236  5.4e+01  0.0e+00  9.6e+01
  237  5.4e+01  1.4e+02  0.0e+00
  238  4.1e+01  0.0e+00  9.3e+01
  239  4.0e+01  1.4e+02  0.0e+00
  240  2.9e+01  0.0e+00  9.3e+01
  241  2.9e+01  1.4e+02  0.0e+00
  242  1.9e+01  0.0e+00  9.2e+01
  243  1.9e+01  1.4e+02  0.0e+00
  244  1.3e+01  0.0e+00  9.1e+01
  245  1.2e+01  1.4e+02  0.0e+00
  246  8.2e+00  0.0e+00  9.4e+01
  247  8.2e+00  1.4e+02  0.0e+00
  248  5.3e+00  0.0e+00  9.3e+01
  249  5.3e+00  1.3e+02  0.0e+00
  250  3.7e+00  0.0e+00  9.7e+01
  251  3.7e+00  1.4e+02  0.0e+00
  252  2.5e+00  0.0e+00  8.9e+01
  253  2.5e+00  1.4e+02  0.0e+00
  254  1.7e+00  0.0e+00  9.5e+01
  255  1.7e+00  1.4e+02  0.0e+00
  256  1.2e+00  0.0e+00  1.0e+02
  257  1.2e+00  1.3e+02  0.0e+00
  258  8.5e-01  0.0e+00  9.9e+01
  259  8.5e-01  1.4e+02  0.0e+00
  260  6.2e-01  0.0e+00  1.0e+02
  261  6.2e-01  1.3e+02  0.0e+00
  262  4.8e-01  0.0e+00  1.1e+02
  263  4.8e-01  1.2e+02  0.0e+00
  264  3.8e-01  0.0e+00  1.0e+02
  265  3.8e-01  1.3e+02  0.0e+00
  266  3.1e-01  0.0e+00  1.0e+02
  267  3.1e-01  1.4e+02  0.0e+00
  268  2.4e-01  0.0e+00  9.7e+01
  269  2.4e-01  1.4e+02  0.0e+00
  270  1.7e-01  0.0e+00  9.0e+01
  271  1.7e-01  1.5e+02  0.0e+00
  272  1.1e-01  0.0e+00  9.1e+01
  273  1.1e-01  1.4e+02  0.0e+00
  274  7.6e-02  0.0e+00  9.9e+01
  275  7.6e-02  1.3e+02  0.0e+00
  276  5.5e-02  0.0e+00  9.8e+01
  277  5.5e-02  1.4e+02  0.0e+00
  278  4.0e-02  0.0e+00  9.9e+01
  279  4.0e-02  1.4e+02  0.0e+00
  280  2.9e-02  0.0e+00  9.6e+01
  281  2.9e-02  1.4e+02  0.0e+00
  282  2.0e-02  0.0e+00  9.5e+01
  283  2.0e-02  1.4e+02  0.0e+00
  284  1.4e-02  0.0e+00  9.3e+01
  285  1.4e-02  1.4e+02  0.0e+00
  286  9.1e-03  0.0e+00  8.8e+01
  287  9.1e-03  1.4e+02  0.0e+00
  288  5.8e-03  0.0e+00  9.2e+01
  289  5.8e-03  1.4e+02  0.0e+00
  290  4.0e-03  0.0e+00  1.0e+02
  291  4.0e-03  1.3e+02  0.0e+00
  292  3.0e-03  0.0e+00  1.0e+02
  293  3.0e-03  1.3e+02  0.0e+00
  294  2.2e-03  0.0e+00  9.9e+01
  295  2.2e-03  1.4e+02  0.0e+00
  296  1.6e-03  0.0e+00  9.5e+01
  297  1.6e-03  1.4e+02  0.0e+00
  298  1.1e-03  0.0e+00  9.3e+01
  299  1.1e-03  1.4e+02  0.0e+00
  300  7.9e-04  0.0e+00  9.7e+01
  301  7.9e-04  1.4e+02  0.0e+00
  302  5.6e-04  0.0e+00  9.9e+01
  303  5.6e-04  1.3e+02  0.0e+00
  304  4.1e-04  0.0e+00  9.9e+01
  305  4.1e-04  1.4e+02  0.0e+00
  306  2.8e-04  0.0e+00  9.1e+01
  307  2.8e-04  1.4e+02  0.0e+00
  308  1.8e-04  0.0e+00  8.9e+01
  309  1.8e-04  1.4e+02  0.0e+00
  310  1.2e-04  0.0e+00  9.2e+01
  311  1.2e-04  1.4e+02  0.0e+00
  312  7.7e-05  0.0e+00  9.2e+01
  313  7.7e-05  1.4e+02  0.0e+00
  314  5.0e-05  0.0e+00  8.8e+01
```

File: src/gpmr.jl [hunk 1]
```diff
@@ -358,8 +358,21 @@ function gpmr!(solver :: GpmrSolver{T,S}, A, B, b :: AbstractVector{T}, c :: Abs
...
         push!(zt, zero(T), zero(T))
       end
-      @. V[k+1] = q / Haux  # hₖ₊₁.ₖvₖ₊₁ = q
-      @. U[k+1] = p / Faux  # fₖ₊₁.ₖuₖ₊₁ = p
+
+      # hₖ₊₁.ₖ ≠ 0
+      if Haux ≠ 0
+        @. V[k+1] = q / Haux  # hₖ₊₁.ₖvₖ₊₁ = q
+      else
+        V[k+1] .= zero(T)  # hₖ₊₁.ₖ = 0 ⇔ vₖ₊₁ = 0
```
Comment (dpo) [inline src/gpmr.jl:366]: Could you please add some comments here to explain what is going on, so we can remember later?

**Review by dpo — COMMENTED** [review]
This is great, thank you. I think there is still some investigation to be done to determine what problem we've solved when h{k+1,k} = 0. After our conversation, I thought that the situation here is not so different from USYMLQR; we also have two solutions that we add together. The difference is that we do it implicitly.

File: src/gpmr.jl (hunk 1 above)
Comment (amontoison) [reply src/gpmr.jl]: Yes, I added some comments about breakdowns.

**Comment by amontoison** [comment]
> This is great, thank you. I think there is still some investigation to be done to determine what problem we've solved when h{k+1,k} = 0. After our conversation, I thought that the situation here is not so different from USYMLQR; we also have two solutions that we add together. The difference is that we do it implicitly.

I investigated this morning and I know what problem is solved when `h_{k+1,k} = 0` or `f_{k+1,k} = 0`.

GPMR implicitly adds the two solutions of BLOCK-GMRES:
```
[ λM   A ] [ x_k^b  x_k^c ] = [ b  0 ]
[  B  μN ] [ y_k^b  y_k^c ]   [ 0  c ]

x_k = x_k^b + x_k^c
y_k = y_k^b + y_k^c
```

If we denote
```
r_k^b = [ b ] - [ λM   A ] [ x_k^b ]
        [ 0 ]   [  B  μN ] [ y_k^b ]   

r_k^c = [ 0 ] - [ λM   A ] [ x_k^c ]
        [ c ]   [  B  μN ] [ y_k^c ]   
```
, we have the following relations
- `k is odd` and `f_{k+1,k} = 0` =>  `r_k^b = 0`;
- `k is odd` and `h_{k+1,k} = 0` =>  `r_k^c = 0`;
- `k is even` and `f_{k+1,k} = 0` =>  `r_k^c = 0`;
- `k is even` and `h_{k+1,k} = 0` =>  `r_k^b = 0`.


I verified numerically these relations with our implementation of Block-GMRES.

File: test/test_utils.jl [hunk 2]
```diff
@@ -331,3 +331,16 @@ function jacobi(A; T=Float64, pd=true)
   P⁻¹ = Diagonal(J)
   return P⁻¹
 end
+
+# Generate a breakdown in the orthogonal tridiagonalization process and the orthogonal Hessenberg reduction process.
```
Comment (amontoison) [inline test/test_utils.jl:335]: - `transpose = false` => `Ac = A * u₁ / γ = 0`
- `transpose = true` => `A'b = A' * v₁ / β = 0`

**Comment by dpo** [comment]
Can you prove those implications? They should go in the paper. Also, do we have f{k+1,k} "small" => one of the residuals is "small"?

**Comment by dpo** [comment]
I don't mean = 0. I mean small.

**Comment by amontoison** [comment]
When the linear system is inconsistent, the subproblem R_k z_k = t_k is also inconsistent and we have an infinite number of solutions that minimize the residual.
It's not possible to compute the minimum-norm solution because it requires a QLP factorization of R_k, but we can avoid some division by zero and set `\zeta_i = 0` to improve the quality of the solution.

**Comment by amontoison** [comment]
> I don't mean = 0. I mean small.

Humm, the underlying problem is :
Does the following matrix is rank-deficient of not ?
```
[ 0          f_{k+1,k} ]
[ h_{k+1,k}  0         ]
```
When `f_{k+1,k} = 0` or `h_{k+1,k} = 0` we can easily answer the question but when their value are small it can be also related to the magnitude of the coefficients of `A` and  `B`.

**Comment by amontoison** [comment]
@geoffroyleconte 
Do you need a new release ?

**Pull request merged**